### PR TITLE
Fix MSPileup APIs

### LINKED ISSUE #11425: [MS] Define MSPileup RESTful endpoint and actions for pileup data
**Impact of the new feature**
 New WMCore microservice

**Is your feature request related to a problem? Please describe.**
Creation of a new microservice to make pileup management more efficient and easier.
Subtask of https://github.com/dmwm/WMCore/issues/9779

**Describe the solution you'd like**
In addition to the `status` and `info` standard REST APIs, we should provide a way to privileged users and systems to be able to create, read, update and delete pileup ids (CRUD operations).

My suggestion is to provide a new `pileups` endpoint, likely under `ms-pileup/data/pileups` and perform an action according to the HTTP method (and possibly data provided) executed, such as:
* GET: provided a pileup id, return its definition in the database (or empty dictionary if none).
* POST: provided an input json object, it creates a new pileup data structure. The structure needs to go under some validation; and the user need to have enough privileges for that.
* PUT: provided an input json object, it updates an already existent pileup data structure. Validation over the input data needs to be performed, and the user needs to have enough privileges for such operation as well. 
* DELETE: provided a single pileup id, it deletes that pileup structure from the database. User (or machine) needs to have super powers to execute it.

Sending and receiving data should be done with JSON objects (Content-Type: application/json).

On what concerns the validation, CRIC roles should be used for the moment. The service configuration file has to support a definition of group/roles and what actions they can perform.

**Describe alternatives you've considered**
The DELETE action should likely be done only by the system, after a grace period of inactivity has passed (after the pileup id has been active:False for some long period of time).

**Additional context**
None

### PR DESCRIPTION
**Pull request by vkuznet**
Fixes #11425 

#### Status
ready

#### Description
Fix several issues with MSPileup HTTP APIs:
- Fix MSPileup query logic, commit 73c3d81e4ee72decc2db4bd244cb5c41cc3fa82d
- remove REST `API` key about API (it is not used by MSPIleup data logic), commit 5828f2252070f168318d791eabbce7ed62f7f36b
- Adjust docstring to represent current end-point, commit bfd09cec9e764efd63805b3660f410eb61a870bc
- Fix `REST/Format.py` module where I defined `obj` used in exception, and add `cherrypy.HTTPError` exception which can be thrown by underlying service, commit e0fdd133fe2008fc6b0981162eb2972884cb234a
- Fix `REST/Server.py` module to add missing `cherrypy.HTTPError` exception, commit 9c5130df871ad0b57c3d30dd8f3fd1d2444fb330
All aforementioned errors were found during validation tests of MSPileup service. In particular, the REST modules error revealed long standing issue of Python exception error, when lower module exception was not propagated to the client since another exception was thrown. In particular here is an example of such multi-thrown exception:
```
# client made a POST HTTP request to MSPileup
curl -X POST -H "Content-type: application/json" -d@./tmp/pileup/bad.json "http://localhost:8249/ms-pileup/data/pileup"

# It received the following error (only shown relevant part of HTTP response):
...
        <h2>500 Internal Server Error</h2>
        <p>Execution error</p>
        <pre id="traceback"></pre>
...

# while on server side it was shown as following:
2023-02-06 18:18:07,134:ERROR:MSPileupObj: pileupType value BAD is neither of ['classic', 'premix']
2023-02-06 18:18:07,134:ERROR:MSPileupData: Failed to create MSPileupObj, MSPileup input is invalid, pileupType value BAD
is neither of ['classic', 'premix']
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/WMCore/MicroService/MSPileup/MSPileupData.py", line 94, in createPileup
    obj = MSPileupObj(pdict, validRSEs=self.validRSEs)
  File "/usr/local/lib/python3.8/site-packages/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py", line 66, in __ini
t__
    raise Exception(msg)
Exception: MSPileup input is invalid, pileupType value BAD is neither of ['classic', 'premix']
2023-02-06 18:18:07,177:ERROR:MSPileupData: {
    "data": {
        "pileupName": "/sdkjfhsdkf/klsjdflksdj/PREMIX",
        "pileupType": "BAD",
        "expectedRSEs": [],
        "currentRSEs": [],
        "fullReplicas": 1,
        "campaigns": [
            "test"
        ],
        "containerFraction": 0.5,
        "replicationGrouping": "ALL",
        "active": true,
        "pileupSizeGB": 123,
        "rulesList": []
    },
    "message": "Failed to create MSPileupObj, MSPileup input is invalid, pileupType value BAD is neither of ['classic', 'premix']",
    "code": 3,
    "error": "MSPileupError"
}
[06/Feb/2023:18:18:07]  SERVER REST ERROR WMCore.REST.Error.ExecutionError 46bde58da62d83ecf9a8aca3e9731dd0 (Execution error)
[06/Feb/2023:18:18:07]    Traceback (most recent call last):
[06/Feb/2023:18:18:07]      File "/usr/local/lib/python3.8/site-packages/WMCore/REST/Format.py", line 251, in stream_chunked
[06/Feb/2023:18:18:07]        for obj in stream:
[06/Feb/2023:18:18:07]      File "/usr/local/lib/python3.8/site-packages/WMCore/MicroService/Service/Data.py", line 165, in post
[06/Feb/2023:18:18:07]        mspileupError(doc)
[06/Feb/2023:18:18:07]      File "/usr/local/lib/python3.8/site-packages/WMCore/MicroService/Service/Data.py", line 73, in mspileupError
[06/Feb/2023:18:18:07]        raise cherrypy.HTTPError(400, msg) from None
[06/Feb/2023:18:18:07]    cherrypy._cperror.HTTPError: (400, "MSPileupError: Failed to create MSPileupObj, MSPileup input is invalid, pileupType value BAD is neither of ['classic', 'premix'], code: 3")
[06/Feb/2023:18:18:07]
[06/Feb/2023:18:18:07]    During handling of the above exception, another exception occurred:
[06/Feb/2023:18:18:07]
[06/Feb/2023:18:18:07]    Traceback (most recent call last):
[06/Feb/2023:18:18:07]      File "/usr/local/lib/python3.8/site-packages/WMCore/REST/Format.py", line 262, in stream_chunked
[06/Feb/2023:18:18:07]        % (obj, type(obj), str(exp)))
[06/Feb/2023:18:18:07]    UnboundLocalError: local variable 'obj' referenced before assignment
[06/Feb/2023:18:18:07]  SERVER HTTP ERROR cherrypy._cperror.HTTPError e69e87f9bdaa36c8802815848c49f0cb ((500, 'Execution error'))
[06/Feb/2023:18:18:07]    Traceback (most recent call last):
[06/Feb/2023:18:18:07]      File "/usr/local/lib/python3.8/site-packages/WMCore/REST/Server.py", line 749, in default
[06/Feb/2023:18:18:07]        return self._call(RESTArgs(list(args), kwargs))
[06/Feb/2023:18:18:07]      File "/usr/local/lib/python3.8/site-packages/WMCore/REST/Server.py", line 865, in _call
[06/Feb/2023:18:18:07]        return stream_maybe_etag(apiobj.get('etag_limit', self.etag_limit), etagger, reply)
[06/Feb/2023:18:18:07]      File "/usr/local/lib/python3.8/site-packages/WMCore/REST/Format.py", line 619, in stream_maybe_etag
[06/Feb/2023:18:18:07]        raise cherrypy.HTTPError(int(err), message)
[06/Feb/2023:18:18:07]    cherrypy._cperror.HTTPError: (500, 'Execution error')
[06/Feb/2023:18:18:07] ms-pileup-595657876c-dj7j5 188.185.65.135 "POST /ms-pileup/data/pileup HTTP/1.1" 500 Internal Server Error [data: 4573 in 745 out 45056 us ] [auth: OK "/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=valya/CN=443502/CN=Valentin Y Kuznetsov" "" ] [ref: "" "curl/7.61.1" ]
```

The issue in this error is two-fold:
- missing object definition in `REST/Format.py` module caused the following error: `UnboundLocalError: local variable 'obj' referenced before assignment`
- while, `REST/Server.py` were missing handling of `cherrypy.HTTPError` exception too which caused its substitution with `cherrypy._cperror.HTTPError: (500, 'Execution error')`error returned to the client

With provided change the actual error returned to the client is this one:
```
...
        <h2>400 Bad Request</h2>
        <p>MSPileupError: Failed to create MSPileupObj, MSPileup input is invalid, pileupType value BAD is neither of ['classic', 'premix'], code: 3</p>
        <pre id="traceback"></pre>
...
```
and we do have proper error in a server log too:
```
2023-02-06 15:48:00,289:ERROR:MSPileupObj: pileupType value BAD is neither of ['classic', 'premix']
2023-02-06 15:48:00,289:ERROR:MSPileupData: Failed to create MSPileupObj, MSPileup input is invalid, pileupType value BAD is neither of ['classic', 'premix']
Traceback (most recent call last):
  File "/Users/vk/CMS/DMWM/GIT/WMCore/src/python/WMCore/MicroService/MSPileup/MSPileupData.py", line 94, in createPileup
    obj = MSPileupObj(pdict, validRSEs=self.validRSEs)
  File "/Users/vk/CMS/DMWM/GIT/WMCore/src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py", line 66, in __init__
    raise Exception(msg)
Exception: MSPileup input is invalid, pileupType value BAD is neither of ['classic', 'premix']
2023-02-06 15:48:00,482:ERROR:MSPileupData: {
    "data": {
        "pileupName": "/sdkjfhsdkf/klsjdflksdj/PREMIX",
        "pileupType": "BAD",
        "expectedRSEs": [],
        "currentRSEs": [],
        "fullReplicas": 1,
        "campaigns": [
            "test"
        ],
        "containerFraction": 0.5,
        "replicationGrouping": "ALL",
        "active": true,
        "pileupSizeGB": 123,
        "rulesList": []
    },
    "message": "Failed to create MSPileupObj, MSPileup input is invalid, pileupType value BAD is neither of ['classic', 'premix']",
    "code": 3,
    "error": "MSPileupError"
}
[06/Feb/2023:15:48:00] vkair 127.0.0.1 "POST /ms-pileup/data/pileup HTTP/1.1" 400 Bad Request [data: 449 in 847 out 195507 us ] [auth: - "" "" ] [ref: "" "curl/7.87.0" ]
```

With all of these fixes now MSPIleup perform correctly the following operations:
- create new (correct) document via HTTP POST
- query existing document via GET HTTP request and via POST HTTP API call
   - for POST HTTP API call we can use the following query
```
# make query spec, e.g. cat spec.json
{"query": {"pileupType": "classic"}}

# query service
curl -X POST -H "Content-type: application/json" -d@./tmp/pileup/spec.json "http://localhost:8249/ms-pileup/data/pileup"
```
- and, if we post bad MSPileup doc, e.g. of the the fields is not validated our client gets proper HTTP error:
```
# place bad json (which contains wrong value for pileupType
curl -X POST -H "Content-type: application/json" -d@./tmp/pileup/bad.json "http://localhost:8249/ms-pileup/data/pileup"

# client will get the following response:
...
        <h2>400 Bad Request</h2>
        <p>MSPileupError: Failed to create MSPileupObj, MSPileup input is invalid, pileupType value BAD is neither of ['classic', 'premix'], code: 3</p>
        <pre id="traceback"></pre>
...
```

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
#11443

### CODE REVIEW

**Comment by vkuznet** [comment]
Alan, sorry for long PR, but I put in separate commits individual issues I found during validation of MSPileup server, one is related in wrong MSPileyp query logic, while other two is related to REST server module issues which unable to properly handle MSPileup `cherrypy.HTTPError` exception. I tried to put as much details in description as possible but fixes to REST module are trivial (they only contains missing exception).

**Comment by amaltaro** [comment]
@vkuznet Valentin, thank you for these details. Changes are looking good, but I need to review it with more attention.

I wanted to make a few notes/requests though:
* the 4th and 5th commits are related to the HTTPError exception in the REST package. So I'd suggest to squash those 2 commits together.
* in general, we should have a GH issue to properly track it. But I think it's okay to refer to the initial GH issue + the detailed description you gave here.
* most of the pull request labels start with "PR:", and we usually don't apply these labels to PRs (only to issues).

Lastly, I am not comfortable with having end users to freely run queries in our backend. So either we:
1) disable that functionality for the moment
2) or perform a very basic data sanitization before executing it in MongoDB.

**Comment by vkuznet** [comment]
Alan, I'll add additional query validation to this PR during today, and we'll decide before production do we need flexible queries to give to dev-teams. In my view, the functionality should be preserved but we do not need to expose/explain it to dev-teams and rather observe their pattern of work and decide later if such flexible queries should be allowed in their workflows.

**Comment by vkuznet** [comment]
Alan, I added necessary changes to support validation of the JSON spec queries. But for that I did the following:
- I move `schema` method of `MSPileupObj` class to be a stand-alone function (as I originally made it)
- then I used `schema` function in validation of input query
   - the validation check that input keys in spec JSON are part of the schema
- I added appropriate new `MSPileupShcemaError` exception
- and, I provided appropriate unit tests for this use-case

I did not squash any commit as it is mixture of different use-cases, in fact I do not want to mess with them, and I think it is better to keep in this particular case everything separate. But if you think otherwise please let me know and I'll squash according to your guidelines.

**Review by amaltaro — CHANGES_REQUESTED** [review]
Valentin, please have a look at the following unit test:
```
Api_t.Tester:test_multi_throw5b changed from success to error
```

which is marked as "error" for the last 2 runs, so probably something legit.

**Comment by vkuznet** [comment]
Alan, it is very mysterious error which so far I can't neither reproduce or even fine. Here is what I found:
- the actual reported error looks like this 
```(25, 'Inappropriate ioctl for device')
-------------------- >> begin captured stdout << ---------------------

    ERROR: 'X-REST-Status':'200' not in headers
    Show: [B]ody [H]eaders [S]tatus [U]RL; [I]gnore, [R]aise, or sys.e[X]it >> 
--------------------- >> end captured stdout << ----------------------
```
- it comes from `test/python/WMCore_t/REST_t/Api_t.py` line 227, where the codebase uses `cheroot.test` module
- `cheroot` is not in `requirements.txt` of WMCore
- I assume it comes from cherrypy since it is part of its repository, see https://github.com/cherrypy/cheroot
- I installed the specific version of cherrypy listed in our `requirements.txt`
- and, yet if I run the test I get complain about missing `main`
```
python3.8 test/python/WMCore_t/REST_t/Api_t.py
Traceback (most recent call last):
  File "test/python/WMCore_t/REST_t/Api_t.py", line 273, in <module>
    webtest.main()
AttributeError: module 'cheroot.test.webtest' has no attribute 'main'
```
- I even tried to run this test codebase by copying it to MS pod and I got similar error from ms-output pod:
```
_reqmgr2ms@ms-output-7d9ddf9c6c-tgbhg:/data$ python Auth_t.py
Traceback (most recent call last):
  File "Auth_t.py", line 273, in <module>
    webtest.main()
AttributeError: module 'cheroot.test.webtest' has no attribute 'main'
```

Based on my findings I have no idea how this particular test runs on Jenkins and I have no idea where `cheroot` comes from and I have no idea if we have a real problem with this test or not. If you know anything about please provide me further details. But from the reading the code of this test it tests REST module where we did added few things in this PR for handling cherrypy HTTP error and in #11462 where we skip authz in local host. So, it may be that this particular tests something in REST and it is missing HTTP 200 response in header, but since it complains about `Inappropriate ioctl for devic` it can be something totally different within Jenkins setup.

Unless I can run this test locally I have no ability to debug it further.

File: src/python/WMCore/MicroService/MSPileup/MSPileupData.py [hunk 1]
```diff
@@ -165,6 +191,16 @@ def getPileup(self, spec, projection=None):
...
         results = []
 
+        # remove API key from the spec as it defines pileup API
+        if isinstance(spec, dict) and 'API' in spec:
+            spec.pop('API')
+
+        self.logger.info("getPileup spec=%s, projection=%s", spec, projection)
+        err = self.validate(spec, projection)
+        if len(err) != 0:
+            self.logger.error("getPileup validation error %s", err)
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSPileup/MSPileupData.py:201]: I'd suggest to rephrase it to something like "Get pileup query didn't pass the validation. Error: %s"

File: src/python/WMCore/MicroService/MSPileup/MSPileupData.py [hunk 2]
```diff
@@ -165,6 +191,16 @@ def getPileup(self, spec, projection=None):
         """
         results = []
 
+        # remove API key from the spec as it defines pileup API
+        if isinstance(spec, dict) and 'API' in spec:
+            spec.pop('API')
+
+        self.logger.info("getPileup spec=%s, projection=%s", spec, projection)
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSPileup/MSPileupData.py:198]: I'd suggest to move this line below the next if statement. Maybe rephrase as well to something like "Getting pileup with spec: %s, projection: %s"

**Review by amaltaro — CHANGES_REQUESTED** [review]
@vkuznet Valentin, this is looking pretty good in general. I left some minor comments along the code that would make it look better IMO.

On what concerns the commits, there is definitely way too many commits for a single PR. I would suggest to squash them as follows:
* 1st commit: contains the first commit and anything else that is not in the next 2 commits.
* 2nd commit: commits 4 and 5 (REST.Format and REST.Server)
* 3rd commit: any unit tests

So, you are not only going to squash commits, but also move them around such that you can properly squash them. Please let me know if you need any help with this. Or feel free to provide a brand new PR if you prefer to.

File: src/python/WMCore/MicroService/MSPileup/MSPileupData.py (hunk 1 above)
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupData.py:201]: done

File: src/python/WMCore/MicroService/MSPileup/MSPileupData.py [hunk 3]
```diff
@@ -154,6 +155,31 @@ def updatePileup(self, doc):
             return [err.error()]
         return []
 
+    def validate(self, spec, projection=None):
```
[the remark this answers is not in the document]
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupData.py:158]: done

File: src/python/WMCore/MicroService/MSPileup/MSPileupData.py (hunk 2 above)
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupData.py:198]: done

**Comment by vkuznet** [comment]
Alan, I applied your feedback and squashed commits. Feel free to have another look.

**Comment by vkuznet** [comment]
To summarize:
- https://github.com/dmwm/WMCore/pull/11475/commits/df86b4aa01e50e442409630c0f8add5f8ff5aee2 provides changes to MSpileup classes (query logic, etc.)
- https://github.com/dmwm/WMCore/pull/11475/commits/9b71ff5b2916e8212e01573d23813260cddb8042 provides changes to REST layer (missing object and exceptions)
- https://github.com/dmwm/WMCore/pull/11475/commits/4798692d4fb718ca7c5001111af995f044287c77 contains fixes for unit tests

**Comment by amaltaro** [comment]
@vkuznet please have a look at the failing unit test:
    WMCore_t.MicroService_t.MSPileup_t.MSPileupData_t.MSPileupTest:testMSPileupQuery was added with status error. Must be fixed

I will see if I can figure anything out of the Api_t test.

**Comment by vkuznet** [comment]
test this please

**Review by amaltaro — APPROVED** [review]
Thanks Valentin. I couldn't figure the Api_t unit test. I will try to figure it out tomorrow, with all these changes in.

**Comment by amaltaro** [comment]
@mapellidario @belforte the 2nd commit in this PR has some minor changes to the REST layer. Given that it now catches the cherrypy.HTTPError (before a generic exception), I am tagging you just in case you see anything strange on the CRAB side. It will be in the next WMCore release (likely 2.1.7rc6).
